### PR TITLE
[extractor/tiktok] Update `_VALID_URL`, add `api_hostname` arg

### LINKED ISSUE #5706: [TikTok] Invalid URL
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.11.11** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

France

### Provide a description that is worded well enough to be understood

I've had trouble downloading TikTok links saying "The read operation timed out". While searching for issues about this i came across #2396 which had a "fix" of spoofing the user agent, which i did. I'm running programmatically and changed the user agent using `yt_dlp.utils.std_headers["User-Agent"] = "facebookexternalhit/1.1"`. This worked fine for a while, and now i'm getting a new error that I can't seem to find a fix for.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['--user-agent', 'facebookexternalhit/1.1', '-vU', 'https://vm.tiktok.com/ZTR45GpSF/']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.11.11 [8b64402] (zip)
[debug] Python 3.10.6 (CPython x86_64 64bit) - Linux-5.15.0-56-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.11.0, certifi-2020.06.20, pyxattr-0.7.2, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1723 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.11.11, Current version: 2022.11.11
yt-dlp is up to date (2022.11.11)
[debug] [vm.tiktok] Extracting URL: https://vm.tiktok.com/ZTR45GpSF/
[vm.tiktok] ZTR45GpSF: Downloading webpage
[debug] [generic] Extracting URL: https://www.tiktok.com/@/video/7106798200794926362?_r=1&_d=secCgYIASAHKAESPgo8A2L6gE3aID26hWdYMUo7Ots2KAQqRwjDG%2B60x%2FlLtJURz%2FRy5GL7fxK7hqI5ZC%2FRnX9lm6YXtN4ajMomGgA%3D&checksum=3c957febf57e591b7143c12855188c90bf89ee566ffa421f0f00fa65f8a6f57f&language=en&preview_pb=0&sec_user_id=MS4wLjABAAAAYXeo2peZSIayb4qlfmzwZA95CealXeyCT-jDYv3fvc9rw9lWjNVvzV0hz4ZZKbCP&share_app_id=1233&share_item_id=7106798200794926362&share_link_id=EA88E06F-AA6C-49C2-A526-62165636810E&source=h5_m&timestamp=1670158079&tt_from=copy&u_code=dga01g482c3bbi&user_id=6912637862613828613&utm_campaign=client_share&utm_medium=ios&utm_source=copy
[generic] 7106798200794926362?_r=1&_d=secCgYIASAHKAESPgo8A2L6gE3aID26hWdYMUo7Ots2KAQqRwjDG+60x/lLtJURz/Ry5GL7fxK7hqI5ZC/RnX9lm6YXtN4ajMomGgA=&checksum=3c957febf57e591b7143c12855188c90bf89ee566ffa421f0f00fa65f8a6f57f&language=en&preview_pb=0&sec_user_id=MS4wLjABAAAAYXeo2peZSIayb4qlfmzwZA95CealXeyCT-jDYv3fvc9rw9lWjNVvzV0hz4ZZKbCP&share_app_id=1233&share_item_id=7106798200794926362&share_link_id=EA88E06F-AA6C-49C2-A526-62165636810E&source=h5_m&timestamp=1670158079&tt_from=copy&u_code=dga01g482c3bbi&user_id=6912637862613828613&utm_campaign=client_share&utm_medium=ios&utm_source=copy: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] 7106798200794926362?_r=1&_d=secCgYIASAHKAESPgo8A2L6gE3aID26hWdYMUo7Ots2KAQqRwjDG+60x/lLtJURz/Ry5GL7fxK7hqI5ZC/RnX9lm6YXtN4ajMomGgA=&checksum=3c957febf57e591b7143c12855188c90bf89ee566ffa421f0f00fa65f8a6f57f&language=en&preview_pb=0&sec_user_id=MS4wLjABAAAAYXeo2peZSIayb4qlfmzwZA95CealXeyCT-jDYv3fvc9rw9lWjNVvzV0hz4ZZKbCP&share_app_id=1233&share_item_id=7106798200794926362&share_link_id=EA88E06F-AA6C-49C2-A526-62165636810E&source=h5_m&timestamp=1670158079&tt_from=copy&u_code=dga01g482c3bbi&user_id=6912637862613828613&utm_campaign=client_share&utm_medium=ios&utm_source=copy: Extracting information
[debug] Looking for embeds
ERROR: Unsupported URL: https://www.tiktok.com/@/video/7106798200794926362?_r=1&_d=secCgYIASAHKAESPgo8A2L6gE3aID26hWdYMUo7Ots2KAQqRwjDG%2B60x%2FlLtJURz%2FRy5GL7fxK7hqI5ZC%2FRnX9lm6YXtN4ajMomGgA%3D&checksum=3c957febf57e591b7143c12855188c90bf89ee566ffa421f0f00fa65f8a6f57f&language=en&preview_pb=0&sec_user_id=MS4wLjABAAAAYXeo2peZSIayb4qlfmzwZA95CealXeyCT-jDYv3fvc9rw9lWjNVvzV0hz4ZZKbCP&share_app_id=1233&share_item_id=7106798200794926362&share_link_id=EA88E06F-AA6C-49C2-A526-62165636810E&source=h5_m&timestamp=1670158079&tt_from=copy&u_code=dga01g482c3bbi&user_id=6912637862613828613&utm_campaign=client_share&utm_medium=ios&utm_source=copy
Traceback (most recent call last):
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1485, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1561, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/yt-dlp/yt_dlp/extractor/common.py", line 674, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/bin/yt-dlp/yt_dlp/extractor/generic.py", line 2721, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.tiktok.com/@/video/7106798200794926362?_r=1&_d=secCgYIASAHKAESPgo8A2L6gE3aID26hWdYMUo7Ots2KAQqRwjDG%2B60x%2FlLtJURz%2FRy5GL7fxK7hqI5ZC%2FRnX9lm6YXtN4ajMomGgA%3D&checksum=3c957febf57e591b7143c12855188c90bf89ee566ffa421f0f00fa65f8a6f57f&language=en&preview_pb=0&sec_user_id=MS4wLjABAAAAYXeo2peZSIayb4qlfmzwZA95CealXeyCT-jDYv3fvc9rw9lWjNVvzV0hz4ZZKbCP&share_app_id=1233&share_item_id=7106798200794926362&share_link_id=EA88E06F-AA6C-49C2-A526-62165636810E&source=h5_m&timestamp=1670158079&tt_from=copy&u_code=dga01g482c3bbi&user_id=6912637862613828613&utm_campaign=client_share&utm_medium=ios&utm_source=copy
```

### PR DESCRIPTION
**Pull request by bashonly**
### Update `_VALID_URL`
TikTok has changed the redirect target locations of the `vm.tiktok.com` and `vt.tiktok.com` shortlinks to completely omit the `user_id`, e.g.:
`https://www.tiktok.com/@/video/7106798200794926362`
instead of
`https://www.tiktok.com/@user_id/video/7106798200794926362`

This URL format does not match `TikTokIE`'s `_VALID_URL` pattern, so the returned `url_result` falls back to the generic extractor. Visting this URL in a browser will redirect to the proper URL with a `user_id`. But for some users, `www.tiktok.com` webpages are unreachable via yt-dlp.

This PR adjusts the `_VALID_URL` to accept URLs without a `user_id`, which should have no impact on extraction since API requests do not require a `user_id`, and when falling back to webpage extraction the webpage URL is constructed with `self._create_url`.

Closes #5706

### Add `api_hostname` extractor-arg

This is a follow-up to c9f5ce511877ae4f22d2eb2f70c3c6edf6c1971d, to ensure we will have a workaround available for the next time TikTok decides to break one of their API hostnames for us.

Usage: `--extractor-args tiktok:api_hostname=api16-normal-c-useast1a.tiktokv.com`

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/tiktok.py [hunk 1]
```diff
@@ -35,6 +35,13 @@ class TikTokBaseIE(InfoExtractor):
     _WEBPAGE_HOST = 'https://www.tiktok.com/'
     QUALITIES = ('360p', '540p', '720p', '1080p')
 
+    def _real_initialize(self):
+        api_hostname = traverse_obj(
+            self._configuration_arg('api_hostname', ie_key=TikTokIE.ie_key()), -1)
+        if api_hostname:
+            self.write_debug(f'Imported API hostname from extractor argument: {api_hostname}')
+            self._API_HOSTNAME = api_hostname
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:43]:
```suggestion
    @property
    def _API_HOSTNAME(self):
        return self._configuration_arg(
            'api_hostname', ['api16-normal-c-useast1a.tiktokv.com'], ie_key=TikTokIE)[0]
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tiktok.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/tiktok.py:43]: Should this be done for `DouyinIE` too? It redefines `_API_HOSTNAME` / would need a different default value, but I think we would want the extractor-arg to work for it as well

Comment (pukkandan) [reply yt_dlp/extractor/tiktok.py:43]: We want the same extractor-arg to work for douyin too? Won't that cause one of these to be always broken whenever the arg is used?

Comment (bashonly) [reply yt_dlp/extractor/tiktok.py:43]: Oh, good point. We could do `ie_key=DouyinIE` for douyin, but I guess it's not necessary

Comment (pukkandan) [reply yt_dlp/extractor/tiktok.py:43]: Yes, we can copy it to Douyin when need arises

**Pull request merged**